feat(ci/arm): add arm64 support for linux bin and linux appimage

rktjmp opened the pull request:
**What**

Add support for Arm64 linux builds (binary in tar.gz and AppImage).

This PR comes with a bit of a refactor to the existing CI system, so I will try to layout the proposed changes and reasons.

Some builds are attached to this post, others are sporadically available at https://github.com/rktjmp/neovim/releases but they may sometimes be amd64 builds labeled as arm64 to speed up testing. The app image is zipped so I can attach it to this post.

[nvim-linux-aarch64.tar.gz](https://github.com/rktjmp/neovim/files/7095386/nvim-linux-aarch64.tar.gz)
[nvim-linux-aarch64.appimage.tar.gz](https://github.com/rktjmp/neovim/files/7095380/nvim-linux-aarch64.appimage.tar.gz)

**How**

Broadly, by emulating Arm64 in QEMU via docker and building in that.

The tar.gz and AppImage for both currently supported architectures (arm64 & amd64) are built in Docker to DRY the release.yml file.

Currently there are no Arm64 GH runners or any ETA, which is why QEMU is needed.

**Docker**

The docker scripts mirror the existing runners, so they use Ubuntu 18.04 as the base image with gcc-11 backported via Ubuntus ToolChain PPA (identical to GH's runner).

The docker files expose relevant configuration variables via `--build-arg`/`ARG` options. `release.yml` uses these to configure the build, so most configuration can still be located outside of the dockerfiles.

The docker files are written in a way that they build and the export the relevant artifacts (tar file or AppDir) via the `-o <dir>` docker build option.

In this case, the dockerfiles function more like a shell script of sorts, you don't interact with any containers. You pass some options and an output directory and it does the work.

**tar.gz**

Arm64 bins *must* be built in docker, so instead of having two separate build scripts I have unified both architectures into one job with a `matrix.strategy`.

The `dockerfile-output-targz` file performs this work, its build instructions are largely transplanted out of the existing linux build job in `release.yml`.

**AppImage**

Again, both platforms are unified under a single `matrix.strategy` job. AppImages are built in two steps:

- Build AppDir in Docker.
- Build AppImage on runner.

The current AppImage solution uses [LinuxDeploy](https://github.com/linuxdeploy/linuxdeploy), which:

- Isn't available for Arm64.
- Uses FUSE (it's an AppImage itself), which can't be used during `docker
  build` and requires special dispensation during `docker run`.

Because of this, we only build the "AppDir" and some auxilary files inside docker, then export those back to the GH runner where we can run LinuxDeploy (which simply bundles some directories and some boiler plate files).

Update: We use linuxdeploy and appimagetool now for aarch64 compat, linuxdeploy can't output non x86 binaries. linuxdeploy can possibly be dropped.

The dockerfile `dockerfile-output-appimage` performs this work and functions similarly to the targz setup.

I had to extract a helper `AppRun` script from `genimage.sh` which is in `scripts/appimage/AppRun`.

IMO it would be nice if it could all be done inside the docker builder but I didn't want to overstep (or manually create the whole app image structure).

**Breaking**

The current PR has some breaking changes:

- I unified exported filenames to be `nvim-linux-<arch>.<tar.gz|appimage>`
  which is a bit easier to keep track of in `release.yml`.
  - This would break scripts/flatpack?/others who expect the old names.
  - These can be re-named in the publish step to match the current release names.

**Todo/RFC**

- [ ] Provide default arguments in dockerfiles? (`CC=gcc-11` unless specified, what about `CMAKE_BUILD_TYPE`?)
- [ ] Should the targz builder be exposed to end users (moved out of `.github/workflows/`)?
- [ ] Is it worth discussing a more coherent naming structure now? (nvim-win-64.zip?)
- [ ] arm64 vs aarch64 (aarch64 seems to be the "new name", arm64 was the old clang name, arm64 and amd64 look pretty similar in lists, so maybe aarch64 is a better choice.
- [ ] Check that "releases" (ie: v.0.6.0) are built and tagged correctly.
- [ ] Check that the app-images actually work on real hardware...
  - [ ] Also check `UPDATE_INFORMATION` env var contents (`nvim-<arch>`?`)
- [ ] Pin QEMU image version
- [ ] Re-enable Windows/Mac build steps (disabled due to slow turn around time).
- [ ] genimage.sh was probably written for an older version of linuxdeploy and has some old env-vars?
  - https://github.com/linuxdeploy/linuxdeploy-plugin-appimage for env details

I am aware that using docker adds some additional complexity by adding a layer of indirection, but:

- It *must* be used for QEMU/Arm64 (until bare runners exist).
- It may give a more portable/stable build platform (re: brew issues on the
  macos runner last month, not directly comparable but perhaps you can avoid
  similar issues in the future on linux?)
- It gives us a user-facing dockerfile that developers can use to build neovim
  in an isolated environment (say someone wants to build but cant install
  tooling, or is having issues building you can point them to a known-working
  build system).

I would hope that by taking a look at the dockerfiles and release.yml, it is not too hard to see how future changes can be added. I personally don't see it as that much of an increase in complexity but I am also in the docker ecosystem to a dergree and may not have great perspective on the issue.

**Other Options**

I couldn't cursorily see any timeline on Arm runners but you could argue to just wait until they're here.

The QEMU build is *not fast*, about 20-30 minutes vs 3-4 native. This time scale doesn't *really* matter for nightly though (and windows already takes 15m to build).

Personally I think the transportable nature of docker builds is a plus, and when native arm runners come you can just drop QEMU but I am obviously biased since I wrote the PR.

rktjmp committed:
feat(ci/arm): add arm64 support for linux bin and linux appimage

rktjmp commented on .github/workflows/release.yml:
```diff
@@ -16,134 +16,206 @@ on:
+                 -f .github/workflows/dockerfile-output-targz \
+                 -o $GITHUB_WORKSPACE/build .
+      -
+        name: Output version and release for GH release notes
```
Would enjoy an `if arch == amd64` since we can't actually run the arm bin and only need to run *something* once. Currently it "harmlessly errors" for arm.

rktjmp commented on .github/workflows/release.yml:
```diff
@@ -16,134 +16,206 @@ on:
-          cd "$GITHUB_WORKSPACE/build/release"
-          tar cfz nvim-linux64.tar.gz nvim-linux64
-      - uses: actions/upload-artifact@v2
+          cd $GITHUB_WORKSPACE/build/
+          tar xzf nvim-linux-${{ matrix.arch }}.tar.gz
+          printf '::set-output name=version::%s\n' "$(./nvim-linux-${{ matrix.arch }}/bin/nvim --version | head -n 3 | sed -z 's/\n/%0A/g')"
+          printf '::set-output name=release::%s\n' "$(./nvim-linux-${{ matrix.arch }}/bin/nvim --version | head -n 1)"
```
Would enjoy an if arch == amd64 since we can't actually run the arm bin and only need to run something once. Currently it "harmlessly errors" for arm.

rktjmp replied:
Fixed

ck3mp3r commented:
Hi, I am doing these tests on an M1 Mac mini. I can confirm that the binary within `nvim-linux-arm64.tar.gz` works on ubuntu:latest.
Also works on a raspberry pi 4 running ubuntu.

The app image throws the following error on both docker on Mac OS and also the raspberry pi 4 natively: bash: ./nvim-linux-arm64.appimage: cannot execute binary file: Exec format error
I'm assuming the latter is amd64 build...

clason commented:
Appimages are for Linux only and don't work on macOS -- independent of the architecture.

rktjmp commented:
It should be arm64 but perhaps LinuxDeploy is mushing something when it packages it. It does have an "ARCH" env-var but the documentation is broadly ["todo"](https://docs.appimage.org/packaging-guide/from-source/linuxdeploy-user-guide.html). 

I've created an issue to track https://github.com/rktjmp/neovim/issues/3 since it might make a lot of noise if it needs some back and forth.

justinmk commented on scripts/app-image-support/AppRun:
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
```
nit: use the same spelling everywhere for appimage-related files, "appimage" not "app-image". Also the directory can just be named `appimage`, "support" is redundant.

rktjmp replied:
Renamed.

justinmk commented on .github/workflows/dockerfile-output-appdir:
```diff
@@ -0,0 +1,72 @@
+# If you run docker build with the `-o <dir>` flag, it will copy the contents of
+# the final stage out to that dir.
+#
+# NOTE: This *does not* export an AppImage, it exports data to create one!
```
even so, it would help discoverability if the file is named `dockerfile-output-appimage`, imo.

rktjmp replied:
Renamed.

justinmk commented:
> * It gives us a user-facing dockerfile that developers can use to build neovim

related: https://github.com/neovim/neovim/issues/1866

mjlbach commented:
I tried to get neovim as an org on dockerhub but it's unfortunately taken

rktjmp committed:
feat(ci/arm): add appimagetool + arch runtime to generate appimages

linuxdeploy doesn't seem to have a way to set/pass AppImage runtimes,
which we need to do since we're potentially building aarch64 on x86_64.

appimagetool has this option but is a lower level tool compared to
linuxdeploy, so for now we retain ld to prepare the AppDir and ait to
actually generate the image.

rktjmp committed:
replace amd64,arm64 with x86_64,aarch64 (mostly makes script easier)

rktjmp committed:
run appimage build step in build/ to avoid name collisions

rktjmp committed:
files -> file

rktjmp committed:
disable zsync upload for now so we can actually run publish

rktjmp committed:
strip outdated linuxdeploy/linuxdeploy-plugin-appimage ENVs

rktjmp committed:
add zsync generation by appimagetool

rktjmp committed:
rename runtime -> appimage-runtime

rktjmp committed:
mark VERSION file for appimage as deprecated? lxdeploy does not observe that ENV

rktjmp committed:
update dockerfile-output-appdir notice to reflect current reality

rktjmp committed:
rename dockerfile-output-appdir to dockerfile-output-appimage per review

rktjmp committed:
rename scripts/app-image-support to scripts/appimage per review

rktjmp committed:
only run set-output when building for host arch

rktjmp commented:
aarch64/arm64 appimage builds should now be functional (can only test in QEMU).

Applied review changes.

Interested in broader yes/no input on "using Docker to build all linux outputs" before continuing any work (no point if it wont be merged). I don't see it as an issue but other people have to be able to maintain it too :+1: 

After that probably worth looking at the naming conventions used. Right now it's `x86_64` and `aarch64` just to make some of the appimage scripting simpler but it's not a huge drama to rename them at the publish step barring a zsync updateinformation quirk.

`arm64` and `amd64` are quite similar looking, so using `aarch64` might avoid erroneous "it wont run" issues when people get the wrong binary, assuming the non-arm binary is not left at `linux64`. My brain says "if one is tagged they should all be tagged" but that's just me. Changing from current names does have consequences downstream.

Doing it now might avoid drama when you need `aarch64` and `riscv64`\* builds later on, or maybe if you start shipping intel and M* macOS builds separately.

Of course, renaming can be discussed in a separate PR, just need guidance on what to call the arm builds for this one.

(Personally I think `x86_64` looks pretty ugly in the distributables :zipper_mouth_face:)

\* one can dream, we can build them now though!

markmartirosian commented:
Tested the aarch64 appimage on AWS `t4g.xlarge`.

Works without any issues.

eggbean commented:
Where can I get the aarch64 AppImage? When I saw that this issue was closed I assumed it would be available on the Releases page.

dundargoc commented:
> Where can I get the aarch64 AppImage? When I saw that this issue was closed I assumed it would be available on the Releases page.

I don't think there is an aarch64 AppImage yet.

dundargoc commented:
@rktjmp may I ask what prompted you to close this PR?

rktjmp commented:
It was running on 5 months stale with no one pushing for a merge, seemed pretty dead in the water. 

Sometimes people are just too nice to say "no thank you". The core team has to _want_ to transition to buildx/containers in CI since in the end they'll bear the maintenance of it, seemed like that isn't a current priority which is 100% fine. I'd rather it not just hang around in PR lists, leading some people on to think it's happening or otherwise not want to discuss it.

Beyond that I don't use m1/arm so my personal investment in doing the work was only to help the Neovim project as a whole, not enough to bang on pots and soap box the change (I do think providing user-build instructions as `<container-runtime> build` is probably net positive).  

Probably it's better served by having the build system actually discussed somewhere between the leads and they can weigh up the value of changing or not and go from there. From (foggy) memory, all that was left to do on this PR was confirm some build flags and file names, it's probably a reasonable starting point.

clason commented:
All very understandable. I think there is (always) interest in improving the build CI workflow; the issue is more that not enough people on the current team are familiar enough with it to assess the impact of this significant change. Having ARM64 binaries (especially for macOS, although there people mostly rely on Homebrew) would be worth a bit of effort, in my personal opinion. Another benefit might be if the dockerfile could be used independently to easily create a personal bundled environment (config, plugins, language servers etc.) -- but this is not something that needs to be discussed here.

@jamessan ?

jamessan commented:
This was on my list of things to review, but I've had very limited time lately.

eggbean commented:
The nvim version available on Debian-based systems for aarch64 is so old that my configuration causes multiple errors on startup (v0.4.4), so I have been hoping for an aarch64 appimage for months and having to use vim instead in the mean time.

lewis6991 commented:
@eggbean are you not able to build from source?

clason commented:
I believe interest in using Docker to build releases has just increased somewhat since Github deprecated the Ubuntu-18.04 images so we had to switch to 20.04, which dropped support for distros with GLIBC <2.29.

(Bonus points if we could also use this for xBSD builds.)

dundargoc commented:
Geez this is a big PR. I'll go through the changes and see which parts of this we can use. I'll be honest: this looks like a lot extra work that we'll need to maintain.

rktjmp commented:
Good luck :smile: 

It's likely so out of sync with the current build system it's not worth it.

 If I were me I'd probably just try to pull the docker files out and sync them with the current build practices then fold those back into the current CI (without all the arm/qemu fluff), then circle back around to arm later? There's cruft there trying to get tools to agree with each other that you could ignore then - also the tools have probably improved/been-replaced.

I think it's easier to view release.yml directly and ignore the diff.

dundargoc commented:
@rktjmp I'm sorry, but I'm gonna have to say no on this one. It's just not a maintainable solution I'm afraid. Don't get me wrong, this is super impressive, which is unfortunately part of the problem :). Still, thank you for this work and sorry that it took this long to give you a real answer.
